html_versions.py

```diff
@@ -0,0 +1,239 @@
+        for module in modules:
+            launcher_version = launcher_versions.get(module, None)
+            cfg_ioc_version = cfg_ioc_versions.get(module, None)
+            try:
+                module_details.append(handle_one_module(module_cfg, module,
+                                                        launcher_version,
+                                                        cfg_ioc_version, area))
+            except AssertionError as e:
+                log.warn('Failed to process module {}: {}'.format(module, e))
+
+    return module_details
+
+
+def render(mod_details):
+    """Create and write HTML report for list of modules.
+
+    Args:
+        List of ModDetails objects
+    """
+    # The largest number of dependencies for any one module
+    max_deps = max(len(details.deps) for details in mod_details)
+    # Sort by module name
+    sorted_details = sorted(mod_details, key=lambda md: md.name)
+    # Script name as module i.e. no trailing .py
+    script_name = __file__.split('.')[0]
+    env = jinja2.Environment(loader=jinja2.PackageLoader(script_name,
+                                                         RESOURCES_DIR))
+    template = env.get_template(HTML_TEMPLATE)
+    header_tags = ['thead', 'tfoot']
+    full_html = template.render(header_tags=header_tags, nheaders=max_deps,
+                                mod_details=sorted_details)
+
+    with open(REPORT, 'w') as f:
+        f.write(full_html)
+    print('Created HTML report: {}'.format(REPORT))
+
+
+def get_launcher_versions(gen_cfg, module_cfg):
+    """Determine where possible versions of modules used in the launcher.
+
+    Returns:
+        dict: module name => version string
+    """
+    mirror_root = gen_cfg.get('general', 'mirror_root')
+    apps_xml = gen_cfg.get('launcher', 'apps_xml')
+    new_apps_xml = gen_cfg.get('launcher', 'new_apps_xml')
+    lxml = launcher.LauncherXml(apps_xml, new_apps_xml)
+    cmds = lxml.get_cmds()
+    cmd_dict = launcher.get_updated_cmds(cmds, module_cfg, mirror_root)
+    launcher_versions = {cmd.module_name: cmd.version for cmd in cmd_dict.keys()}
+    return launcher_versions
+
+
+def get_configure_ioc_versions(ioc_names):
+    """Determine where possible versions of modules used in configure-ioc.
+
+    Returns:
+        dict: module name => version string or 'work'
+    """
+    cfg_ioc = subprocess.check_output(CFG_IOC_CMD).strip().split('\n')
+    ioc_paths = [path for _, path in (line.split() for line in cfg_ioc)]
+    versions = {}
+    for ioc_name in ioc_names:
+        for ioc_path in ioc_paths:
+            if ioc_name in ioc_path:
+                if '/work/' in ioc_path:
+                    versions[ioc_name] = 'work'
+                else:
+                    index = ioc_path.index(ioc_name)
+                    end = ioc_path[index+len(ioc_name)+1:]
+                    version = end.split(os.path.sep)[0]
+                    versions[ioc_name] = version
```

Comment:
not really any need to declare the 'version' variable here; this might slow down loop execution too